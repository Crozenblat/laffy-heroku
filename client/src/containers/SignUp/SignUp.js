import React, {useState} from "react";
import styled from "styled-components";
import {useDispatch} from "react-redux";
import {useHistory} from "react-router-dom";
import {authLoginUser} from "../../store/features/authorizationSlice";
import axios from "axios";

import PasswordVisibilityToggle from "../../components/UI/Buttons/PasswordVisibilityToggle";
import SubmitButton from "../../components/UI/Buttons/SubmitButton";

import useCreateFormInputs from "../../hooks/useCreateFormInputs"
import useFormInputChangedHandler from "../../hooks/useFormInputChangedHandler";
import useTogglePasswordVisibility from "../../hooks/useTogglePasswordVisibility";

const SgnUpFrm = styled.form`

`;

const SbmtBtnPosWrppr = styled.span`
    display: flex;
    justify-content: center;
`;

const SignUpForm = props => {
    const [controls, setControls] = useState({
        signUpUsername:{
            label: "Username",
            value: "",
            type: "text",
            isValid: false,
            errMsg: null,
            touched: false,
            validationRules:{
                required: true,
            }
        },
        signUpPassword:{
            label: "Password",
            value: "",
            type: "password",
            isValid: false,
            errMsg: null,
            touched: false,
            validationRules:{
                required: true,
                minLength: 8
            }
        }
    });

    const dispatch = useDispatch();
    const history = useHistory();

    const inputChangedHandler = useFormInputChangedHandler;

    const formInputs = useCreateFormInputs(controls, setControls, inputChangedHandler);

    const togglePasswordVisibility = useTogglePasswordVisibility;

    const onSubmitHandler = async (event) => {
        event.preventDefault();

        try{
            await axios.post("/signup", {username: controls.signUpUsername.value, password: controls.signUpPassword.value});
            let login = await axios.post("/login", {username: controls.signUpUsername.value, password: controls.signUpPassword.value}, {withCredentials: true});
            dispatch(authLoginUser({userId: login.data.userId, username: login.data.username}));
            history.push("/");
        }catch(err){
            console.log(err);
            setControls(prev => {
                return {
                    ...prev,
                    [err.response.data.input]:{
                        ...prev[err.response.data.input],
                        errMsg: err.response.data.message
                    }
                };
            });
        };
    };

    return(
        <SgnUpFrm>
            {formInputs}
            <PasswordVisibilityToggle click={() => togglePasswordVisibility(controls, setControls)} id="signUpShowPassword">Show Password</PasswordVisibilityToggle>
            <SbmtBtnPosWrppr>
                <SubmitButton click={onSubmitHandler} disabled={!controls.signUpUsername.isValid || !controls.signUpPassword.isValid}>Sign Up</SubmitButton>
            </SbmtBtnPosWrppr>
        </SgnUpFrm>
    );
};

export default SignUpForm;