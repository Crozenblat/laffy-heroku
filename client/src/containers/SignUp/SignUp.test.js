import React from "react";
import {render, screen} from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import {BrowserRouter} from "react-router-dom";

import seed from "../../../backend/seed";

import SignUp from "./SignUp";

describe("Sign Up Form", () => {
    const userOne = seed.users[0];
    let usernameField;
    let passwordField;
    let signUpBtn;

    beforeEach(() => {
        render(
            <BrowserRouter>
                <SignUp/>
            </BrowserRouter>
        );

        usernameField = screen.getByTestId("signup-username-field");
        passwordField = screen.getByTestId("signup-password-field");
        signUpBtn = screen.getByRole("button", {name: "Sign Up"});
    });

    it("should render correctly", () => {
        let wrapper = render(
            <BrowserRouter>
                <SignUp/>
            </BrowserRouter>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it("should disable the 'sign up' button if one or more of the fields is empty", () => {
        expect(usernameField).toBeInvalid();
        expect(passwordField).toBeInvalid();
        expect(signUpBtn).toBeDisabled();
    });

    it("should display a loading animation whilst attempting to sign up", () => {
        userEvent.type(usernameField, "nonExistentUsername");
        userEvent.type(passwordField, userOne.password);

        eexpect(usernameField).toBeValid();
        expect(passwordField).toBeValid();

        userEvent.click(signUpBtn);
        let loadingAnimation = screen.getByTestId("loading-animation");

        expect(loadingAnimation).toBeInTheDocument();
    });

    it("should accept a unique username and a password at least 8 characters long", async () => {
        userEvent.type(usernameField, "uniqueUsername");
        userEvent.type(passwordField, "newPassword");

        userEvent.click(signUpBtn);
        let loadingAnimation = screen.getByTestId("loading-animation");

        expect(usernameField).toBeValid();
        expect(passwordField).toBeValid();
        await waitFor(() => expect(signUpBtn).toHaveClass("success"));
        expect(loadingAnimation).not.toBeInTheDocument();

    });

    it("should display an error message if the username is already taken", async () => {
        userEvent.type(usernameField, userOne.username);
        userEvent.type(passwordField, "newPassword");

        userEvent.click(signUpBtn);
        let loadingAnimation = screen.getByTestId("loading-animation");

        expect(usernameField).toBeValid();
        expect(passwordField).toBeValid();
        expect(await screen.findByText("Username already taken"));
        expect(signUpBtn).toHaveClass("failure");
        expect(loadingAnimation).not.toBeInTheDocument();
    });

    it("should display an error message if the password is less than 8 characters", async () => {
        userEvent.type(usernameField, "uniqueUsername");
        userEvent.type(passwordField, "short");

        userEvent.click(signUpBtn);

        expect(usernameField).toBeValid();
        expect(passwordField).toBeInvalid();
        expect(await screen.findByText("Password must be at least 8 characters"));
        await waitFor(() => expect(signUpBtn).toHaveClass("failure"));
    });
});