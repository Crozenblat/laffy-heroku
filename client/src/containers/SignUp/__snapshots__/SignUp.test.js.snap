// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up Form should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="sc-Axmtr gyRQxU"
      >
        <div
          class="sc-AxhUy kRznDN"
        >
          <label
            class="sc-AxhCb iIHvLx"
          >
            Username
          </label>
          <input
            class="sc-AxiKw gzpttR"
            placeholder="Username"
            required=""
          />
          <span
            class="sc-AxjAm cjbuaO"
          />
          <p
            class="sc-AxirZ jlGEmD"
          >
            Error Message
          </p>
        </div>
        <div
          class="sc-AxhUy kRznDN"
        >
          <label
            class="sc-AxhCb iIHvLx"
          >
            Password
          </label>
          <input
            class="sc-AxiKw gzpttR"
            min="8"
            placeholder="Password"
            required=""
          />
          <span
            class="sc-AxjAm cjbuaO"
          />
          <p
            class="sc-AxirZ jlGEmD"
          >
            Error Message
          </p>
        </div>
        <span
          class="sc-AxmLO hspwMb"
        >
          <button
            class="sc-AxheI fEiGtH"
            type="form"
          >
            <a
              class="sc-AxgMl bsNeod"
              href=""
            >
              Sign Up
            </a>
          </button>
        </span>
      </form>
    </div>
    <div>
      <form
        class="sc-Axmtr gyRQxU"
      >
        <div
          class="sc-AxhUy kRznDN"
        >
          <label
            class="sc-AxhCb iIHvLx"
          >
            Username
          </label>
          <input
            class="sc-AxiKw gzpttR"
            placeholder="Username"
            required=""
          />
          <span
            class="sc-AxjAm cjbuaO"
          />
          <p
            class="sc-AxirZ jlGEmD"
          >
            Error Message
          </p>
        </div>
        <div
          class="sc-AxhUy kRznDN"
        >
          <label
            class="sc-AxhCb iIHvLx"
          >
            Password
          </label>
          <input
            class="sc-AxiKw gzpttR"
            min="8"
            placeholder="Password"
            required=""
          />
          <span
            class="sc-AxjAm cjbuaO"
          />
          <p
            class="sc-AxirZ jlGEmD"
          >
            Error Message
          </p>
        </div>
        <span
          class="sc-AxmLO hspwMb"
        >
          <button
            class="sc-AxheI fEiGtH"
            type="form"
          >
            <a
              class="sc-AxgMl bsNeod"
              href=""
            >
              Sign Up
            </a>
          </button>
        </span>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-Axmtr gyRQxU"
    >
      <div
        class="sc-AxhUy kRznDN"
      >
        <label
          class="sc-AxhCb iIHvLx"
        >
          Username
        </label>
        <input
          class="sc-AxiKw gzpttR"
          placeholder="Username"
          required=""
        />
        <span
          class="sc-AxjAm cjbuaO"
        />
        <p
          class="sc-AxirZ jlGEmD"
        >
          Error Message
        </p>
      </div>
      <div
        class="sc-AxhUy kRznDN"
      >
        <label
          class="sc-AxhCb iIHvLx"
        >
          Password
        </label>
        <input
          class="sc-AxiKw gzpttR"
          min="8"
          placeholder="Password"
          required=""
        />
        <span
          class="sc-AxjAm cjbuaO"
        />
        <p
          class="sc-AxirZ jlGEmD"
        >
          Error Message
        </p>
      </div>
      <span
        class="sc-AxmLO hspwMb"
      >
        <button
          class="sc-AxheI fEiGtH"
          type="form"
        >
          <a
            class="sc-AxgMl bsNeod"
            href=""
          >
            Sign Up
          </a>
        </button>
      </span>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
